pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -361,12 +352,24 @@ public Representation post(Representation entity) {
           // list of FileItems
           items = upload.parseRequest(getRequest());
 
-          for (final Iterator<FileItem> it = items.iterator(); it.hasNext() && !found;) {
-            final FileItem fi = it.next();
-            if (fi.getFieldName() != null) {
-              found = true;
-              dataFile = new File(tempDir, fi.getFieldName());
-              fi.write(dataFile);
+          for (FileItem fileItem : items) {
+            if (!found) {
+              if (fileItem.getFieldName() != null) {
+                found = true;
+                dataFile = new File(tempDir, fileItem.getFieldName());
+                fileItem.write(dataFile);
+              }
+            } else {
+              LOGGER.warn("Got extra file item while pushing segments: " + fileItem.getFieldName());
+            }
+
+            // TODO: remove the try-catch after verifying it will not throw any exception
+            try {
+              // Remove the temp file
+              // When the file is copied to instead of renamed to the new file, the temp file might be left in the dir
+              fileItem.delete();
+            } catch (Exception e) {
+              LOGGER.error("Caught exception while deleting the temp file, should not reach here", e);
```

Comment:
If delete throws exception (which should not happen), get file name from fileItem might throw another exception which break the file upload.